text/0502_aws-vpclattice.md

```diff
@@ -0,0 +1,1070 @@
+# vpcLattice L2 Construct
+
+- [Project Information](#project-information)
+- [Example Impleentation](#example-implementation)
+- [API Design](#proposed-api-design-for-vpclattice)
+- [FAQ](#faq)
+- [Acceptance](#acceptance)
+
+---
+
+## Project Information
+
+### Status
+
+(DRAFT)
+
+### Original Author(s)
+
+@mrpackethead, , @taylaand, @nbaillie
+
+### Tracking Issue
+
+502
+
+### API Bar Raiser
+
+@TheRealAmazonKendra
+
+## Public Issues
+
+( aws-cdk)(vpclattice): L2 for Amazon VPC Lattice #25452
+
+### Prototype Code
```

Comment:
### Service Construct


**allowUnauthenticatedAccess → requireAuthentication**

This setting could be the on/off switch for AuthN on the service

I would invert this setting to requireAuthentication  just because if you set this option you would expect we’re running the AuthN step - and unauthenticated would no longer be possible. If you had set this setting, then any SIGv4 caller could call - unless restrictions were added with the above methods 


**allowExternalPrincipals →  onlyOrganizationIds & onlyAccountIds**

This setting could add conditions to AuthZ policy 
We could strengthen what control can be expressed here by having this be a list of allowed Organization Ids rather than a true/false. 

Setting this to false will trigger the CDK to call the SDK to get the current users Organization Id, that’s not always the correct Organization Id - which Organization someone would want to restrict things to can depend on how they’re structuring their AWS resources. There could even be multiple organizations which are in use.

The word External in the name of the option could be a little overloaded - so could rename to something like onlyOrganizationIds which will be less ambiguous. If unset or the empty list, we’d apply no organization restriction, so that’s why I suggested only vs allow as a prefix.

Along the same lines onlyAccountIds could be a useful complementary function, that takes a list of account ids to restrict access to. This will also help prevent the caller from being surprised they need to have permission to describe organizations.

(The semantics are a bit different from RAM because we have Anonymous identities)

**allowAnonymousPrincipals →  true / false (default false)**

This setting can add conditions to allow or explicitly DENY Anonymous principal.

If false, then an explicit DENY AuthZ policy should be added to the Service which prevents the principal named “ANONYMOUS”. This will ensure AuthN is not skipped by sending an Authorization: ANONYMOUS header - even when the service owner has used requireAuthentication

This prefix I kept as Allow because this is an option - callers can still be authenticated, but they can also be ANONYMOUS when this is set.

### Service Network Construct

This construct should have security methods and semantics that are identical to what we discuss above for the Service Construct.

There is a bug on line 240 - 
https://github.com/raindancers/aws-cdk/blob/mrpackethead/aws-vpclattice-alpha/packages/%40aws-cdk/aws-vpclattice-alpha/lib/servicenetwork.ts#L240

    const statement = new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      actions: ['vpc-lattice-svcs:Invoke'],
      resources: ['*'],
      principals: [new iam.AnyPrincipal()],
    });

Don’t add this implicit ALLOW * permission - it will allow all Orgs and it will allow the Anonymous principal to call by default. It is better to have a default DENY. Instead make caller explicitly tell us the permissions - that’s important for the Policy configuration so that there are no surprises.

For the share() method , would use only the principals field - you can put into that field the exact lists of principals and org ids that we would have collected above. So no need to set allowExternalPrincipals in the resource share.

https://github.com/raindancers/aws-cdk/blob/mrpackethead/aws-vpclattice-alpha/packages/%40aws-cdk/aws-vpclattice-alpha/lib/servicenetwork.ts#L345C1-L353C1
```
  public share(props: ShareServiceNetworkProps): void {
    new ram.CfnResourceShare(this, 'ServiceNetworkShare', {
      name: props.name,
      resourceArns: [this.serviceNetworkArn],
      allowExternalPrincipals: props.allowExternalPrincipals,
      principals: props.principals,
    });
  }
```
Similar comment for shareToAccounts() on service:

https://github.com/raindancers/aws-cdk/blob/mrpackethead/aws-vpclattice-alpha/packages/%40aws-cdk/aws-vpclattice-alpha/lib/service.ts#L299


### Listener Construct

The listener construct has the ability to set Policy, but it can only set it to the rules that were set on the service.

I would remove this - the auth rules getting generated are granular - but they don’t produce an AuthZ result that is different from what you have achieved setting policy at the Service or Service Network level on Resource ‘*’. You’d need to have even more options to vary some of the policy statements from the API structure to achieve a different AuthZ result.

This could also lead to large policies and push against the upper bound for individual policies, since we auto add new large statements with each sort of match. 

Listeners who need to have fine grained auto at the level of route should provide a full policy statement, rather than our trying to generate from the shape of their API. This will help users make explicit decisions about security, rather than our generating default rules for them - and it will be easier for them to manage these rules since they’d be thier own explicit artifact. 



### Associate VPC Construct

https://github.com/raindancers/aws-cdk/blob/mrpackethead/aws-vpclattice-alpha/packages/%40aws-cdk/aws-vpclattice-alpha/lib/servicenetwork.ts#L391C1-L394C9

Would make the port configurable - this only allows ingress via 443, but that may not be the port in use & protocol may not be HTTPS.

      securityGroup.addIngressRule(
        ec2.Peer.ipv4(props.vpc.vpcCidrBlock),
        ec2.Port.tcp(443),
      );

I’m not so sure I would try to create the SecurityGroup on behalf of customers at all - that’s a resource we’re taking as an input to Lattice - there’s lots of options and considerations when it comes to how users would use these. 

Would the VPC L2 construct cover this sort of convenience?



### HTTP Method Enum

Could include all the VERBs here like OPTIONS, PATCH, TRACE, ...
https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods

Reply:
> ### Service Construct
> **allowUnauthenticatedAccess → requireAuthentication**
> 
> This setting could be the on/off switch for AuthN on the service
> 
> I would invert this setting to requireAuthentication just because if you set this option you would expect we’re running the AuthN step - and unauthenticated would no longer be possible. If you had set this setting, then any SIGv4 caller could call - unless restrictions were added with the above methods

I would by default always pick the more 'secure' default.   So you have to opt out of security. Its a trivial change to invert the logic, and I'm ok with it, it really does'tn change the functionality. 

> 
> **allowExternalPrincipals → onlyOrganizationIds & onlyAccountIds**
> 
> This setting could add conditions to AuthZ policy We could strengthen what control can be expressed here by having this be a list of allowed Organization Ids rather than a true/false.

The grantAccess() method on the Network Class, provides the mechanism to add a list of principals, but its not setting a condition in an iam policy   i'm going to have to go and look for an example if there is a grant like method for conditons.  I dont' remember seeing one.   
The idea of being able to provide a list of principals to create the condition is a good one. I' Need to think about how we should implement it in a cdk-esqe way. 
``
> 
> Setting this to false will trigger the CDK to call the SDK to get the current users Organization Id, that’s not always the correct Organization Id - which Organization someone would want to restrict things to can depend on how they’re structuring their AWS resources. There could even be multiple organizations which are in use.

That was the intent. ( and its currently the default ).  if you do want to allow externals, I think you shoudl be setting this to true,  so you've explicity allowed it.  THat would result in NO condition be added to the policy. 


>
> The word External in the name of the option could be a little overloaded - so could rename to something like onlyOrganizationIds which will be less ambiguous. If unset or the empty list, we’d apply no organization restriction, so that’s why I suggested only vs allow as a prefix.
>

Maybe 'External' is overloaded. but onlyOnlyOrganizationId is confusing.    The intent here is to add a conditon that is restricted to the org that the account is deployed to.  
 
> Along the same lines onlyAccountIds could be a useful complementary function, that takes a list of account ids to restrict access to. This will also help prevent the caller from being surprised they need to have permission to describe organizations.

I'll go and find what the 'grant' like CDK method for a condition is. 


> 
> (The semantics are a bit different from RAM because we have Anonymous identities)
> 
> **allowAnonymousPrincipals → true / false (default false)**
> 
> This setting can add conditions to allow or explicitly DENY Anonymous principal.
> 
> If false, then an explicit DENY AuthZ policy should be added to the Service which prevents the principal named “ANONYMOUS”. This will ensure AuthN is not skipped by sending an Authorization: ANONYMOUS header - even when the service owner has used requireAuthentication

> 
> This prefix I kept as Allow because this is an option - callers can still be authenticated, but they can also be ANONYMOUS when this is set.


 
> ### Service Network Construct
> This construct should have security methods and semantics that are identical to what we discuss above for the Service Construct.


> 
> There is a bug on line 240 - https://github.com/raindancers/aws-cdk/blob/mrpackethead/aws-vpclattice-alpha/packages/%40aws-cdk/aws-vpclattice-alpha/lib/servicenetwork.ts#L240
> 
> ```
> const statement = new iam.PolicyStatement({
>   effect: iam.Effect.ALLOW,
>   actions: ['vpc-lattice-svcs:Invoke'],
>   resources: ['*'],
>   principals: [new iam.AnyPrincipal()],
> });
> ```
> 
> Don’t add this implicit ALLOW * permission - it will allow all Orgs and it will allow the Anonymous principal to call by default. It is better to have a default DENY. Instead make caller explicitly tell us the permissions - that’s important for the Policy configuration so that there are no surprises.
> 
> For the share() method , would use only the principals field - you can put into that field the exact lists of principals and org ids that we would have collected above. So no need to set allowExternalPrincipals in the resource share.
> 
> https://github.com/raindancers/aws-cdk/blob/mrpackethead/aws-vpclattice-alpha/packages/%40aws-cdk/aws-vpclattice-alpha/lib/servicenetwork.ts#L345C1-L353C1
> 
> ```
>   public share(props: ShareServiceNetworkProps): void {
>     new ram.CfnResourceShare(this, 'ServiceNetworkShare', {
>       name: props.name,
>       resourceArns: [this.serviceNetworkArn],
>       allowExternalPrincipals: props.allowExternalPrincipals,
>       principals: props.principals,
>     });
>   }
> ```
> 
> Similar comment for shareToAccounts() on service:
> 
> https://github.com/raindancers/aws-cdk/blob/mrpackethead/aws-vpclattice-alpha/packages/%40aws-cdk/aws-vpclattice-alpha/lib/service.ts#L299
> 



> ### Listener Construct
> The listener construct has the ability to set Policy, but it can only set it to the rules that were set on the service.

The principals in the listener Rule are optional.  If you dont' supply them, it wont' add statemetns to the policy.. However if you do, it will. 

You have the ability to add what ever policy statements you like with the addStatementToPolicy() method.  

For the implmenetations i've done, this has been a very easy and quick way to get something working. As written its provided the ability to have an 'easy' way. and you've got a way to effectively hand build whatever you like..  

 
> I would remove this - the auth rules getting generated are granular - but they don’t produce an >AuthZ result that is different from what you have achieved setting policy at the Service or >Service Network level on Resource ‘*’.

Having best of both worlds here. is a good thing.  CDK is about abstracting the complexity away for the 80%, and providing options for the 20% who dare to do things the hard way. 

>ou’d need to have even more options to vary some of the policy statements from the API structure ?to achieve a different AuthZ result.

Which we have. 

> 
> This could also lead to large policies and push against the upper bound for individual policies, since we auto add new large statements with each sort of match

We can check for policy size at synthtime. What is the maximum policy size permitted?  
.

> Listeners who need to have fine grained auto at the level of route should provide a full policy statement, rather than our trying to generate from the shape of their API. This will help users make explicit decisions about security, rather than our generating default rules for them - and it will be easier for them to manage these rules since they’d be thier own explicit artifact.

Which you can. Again, this is the point of CDK.  Abstractions to make it simple. If you want to have that degree of constrol, nothing stops you here, at the L2 Contruct, or even you can drop down to the L1 Cfn types. 


> ### Associate VPC Construct
> https://github.com/raindancers/aws-cdk/blob/mrpackethead/aws-vpclattice-alpha/packages/%40aws-cdk/aws-vpclattice-alpha/lib/servicenetwork.ts#L391C1-L394C9
> 
> Would make the port configurable - this only allows ingress via 443, but that may not be the port in use & protocol may not be HTTPS.
> 
> ```
>   securityGroup.addIngressRule(
>     ec2.Peer.ipv4(props.vpc.vpcCidrBlock),
>     ec2.Port.tcp(443),
>   );
> ```

This only gets added if a security Group if one is not supplied. As writen the props for AssociateVPC take a security group.

Options:
(a) Default to tcp/443 if you dont' supply one. 
(b) Default to not providing a security group,  ( this seems very bad idea ) if you dont' provide one
(c) make the security group a required prop. 

Its just my experience here, but Id have thought that MOST people will be using tcp 443, so thats what you default to, and leave the option for someone to provide one.    

We are also exposing securityGroups as an attribute, so theres another avenue open for people. 


> I’m not so sure I would try to create the SecurityGroup on behalf of customers at all - that’s >a resource we’re taking as an input to Lattice - there’s lots of options and considerations when >it comes to how users would use these.

If you were doing this in CF, that woudl be the approach..  CDK is all about making it easy. 

> Would the VPC L2 construct cover this sort of convenience?

yes. there are plenty of CDK examples of this. 

> 
> ### HTTP Method Enum
> Could include all the VERBs here like OPTIONS, PATCH, TRACE, ... https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods

Yes, that is definatly a good idea.